File: algorithms/easy/sum_of_values_at_indices_with_k_set_bits.ex
defmodule Solution do
  use Bitwise

  @spec sum_indices_with_k_set_bits(nums :: [integer], k :: integer) :: integer
  def sum_indices_with_k_set_bits(nums, k) do
    nums
    |> Enum.with_index()
    |> Enum.reduce(0, fn {n, i}, acc ->
      if count_one(i, 0) == k do
        acc + n
      else
        acc
      end
    end)
  end

  def count_one(n, ct) when n == 0, do: ct

  def count_one(n, ct) do
    if (n &&& 1) == 1 do
      count_one(n >>> 1, ct + 1)
    else
      count_one(n >>> 1, ct)
    end
  end
end

File: algorithms/easy/sum_of_values_at_indices_with_k_set_bits.rs
fn count(mut i: i32) -> i32 {
    let mut ct = 0;
    while i > 0 {
        ct += i % 2;
        i /= 2;
    }
    ct
}

impl Solution {
    pub fn sum_indices_with_k_set_bits(nums: Vec<i32>, k: i32) -> i32 {
        let mut sum = 0;
        for (i, n) in nums.into_iter().enumerate() {
            if count(i as i32) == k {
                sum += n;
            }
        }
        sum
    }
}
